### PR TITLE
[orc8r] Document and clean up Orc8r e2e security

### PR DESCRIPTION
**Pull request by hcgatewood**
## Summary

Document Orc8r e2e security, to support onboarding external developers, debugging operator issues, and dev alignment on e2e codepath.

In the process of understanding the security codepath, I found and added a number of small upgrades.

Screen Shot 2021-03-05 at 7 32 55 PM

Screen Shot 2021-03-05 at 7 32 59 PM

## Test Plan

- [x] Local sanity check with Docker

## Additional Information

- [ ] This change is backwards-breaking

### CODE REVIEW

File: orc8r/tools/scripts/consolidate_protos.sh [hunk 1]
```diff
@@ -5,17 +5,21 @@
```
Comment (karthiksubraveti) [inline orc8r/tools/scripts/consolidate_protos.sh:5]: What are these scripts used for ?

File: orc8r/gateway/python/scripts/checkin_cli.py [hunk 2]
```diff
@@ -42,51 +44,62 @@ def main():
...
             """,
         'direct_checkin':
             """
-            - Verify state service is running on cloud VM.
-            - Double check the gateway has been registered with the cloud. You
-              can check by going to cloud swagger,
-              (EG. https://127.0.0.1:9443/apidocs/v1/#/), and query the list
-              gateways endpoint.
-            - Check logs for more information (sudo tail -f /var/log/syslog).
+            - Verify cloud's state service is healthy
```
Comment (karthiksubraveti) [inline orc8r/gateway/python/scripts/checkin_cli.py]: Provide instructions on how to check cloud state service is healthy.

File: orc8r/cloud/go/tools/accessc/handlers/common.go [hunk 3]
```diff
@@ -31,11 +31,7 @@ var (
 
 // BuildACLForEntities builds ACL from networks, operators & gateways command
 // line flags
-func BuildACLForEntities(
-	networks,
-	operators,
-	gateways Entities) []*accessprotos.AccessControl_Entity {
-
+func BuildACLForEntities(networks, operators, gateways Entities) []*accessprotos.AccessControl_Entity {
```
Comment (karthiksubraveti) [inline orc8r/cloud/go/tools/accessc/handlers/common.go:34]: minor nit: I would generally prefer making stylistic changes only when we change the actual code. Seems a little distracting from the actual set of changes.

File: orc8r/cloud/go/service/middleware/unary/identity_decorator.go [hunk 4]
```diff
@@ -200,10 +197,7 @@ func findGatewayIdentity(serialNumber string, md metadata.MD) (*protos.Identity,
 	expSeconds := expiration.Unix()
 	// Check if it's Gateway identity
 	if gwIdentity == nil {
-		glog.Infof(
-			"Identity (%s) of Cert SN %s from metadata %+v is not a Gateway",
-			id.HashString(), serialNumber, md)
-
+		glog.Infof("Identity (%s) of Cert SN %s from metadata %+v is not a Gateway", id.HashString(), serialNumber, md)
```
Comment (karthiksubraveti) [inline orc8r/cloud/go/service/middleware/unary/identity_decorator.go:197]: Actually leaving a review for this piece of code prior to this change.

```
	certInfo, err := getCertifierIinfo(serialNumber, md)
	id := certInfo.GetId()
	if err != nil {
		return id, 0, err
```
We should do the error check prior to getting the ID here.

File: orc8r/cloud/docker/nginx/templates/nginx.conf.j2 [hunk 5]
```diff
@@ -148,6 +145,10 @@ http {
     listen 9443 ssl default_server;
     server_name _;
 
+    # Automatically upgrade HTTP requests to HTTPS
```
Comment (karthiksubraveti) [inline orc8r/cloud/docker/nginx/templates/nginx.conf.j2:148]: Thanks for this change.

File: docs/readmes/orc8r/architecture_security.md [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+
+If the bootstrap process is successful, Orc8r signs and returns the CSR from the `RequestSign` request, which the gateway can now use as
+its session certificate. Orc8r enforces a max validity period, so gateways must periodically re-bootstrap to receive updated session
+certificates.
+
+Operators can use the `checkin_cli.py` script to debug issues with this process of checking in and bootstrapping.
+
+## Application-layer access control
+
+Once a request passess TLS termination at the Orc8r proxy, Orc8r also supports finer-grained application-level access control. In Orc8r,
```
Comment (karthiksubraveti) [inline docs/readmes/orc8r/architecture_security.md]: minor nit: passess -> passes

File: docs/readmes/orc8r/architecture_security.md [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+an ACL defines a mapping between identities, where each mapping includes read/write permissions. There are 3 types of identities
```
Comment (karthiksubraveti) [inline docs/readmes/orc8r/architecture_security.md:61]: Can you also provide some code references to the ACL reference being made here towards the end of the document. Will help a curious reader to go and learn about the code better.

File: docs/readmes/orc8r/dev_security.md [hunk 8]
```diff
@@ -0,0 +1,235 @@
...
+Utilized tools
+
+- [`curl`](https://curl.se/)
+- [`grpcurl`](https://github.com/fullstorydev/grpcurl)
+- [`s_client`](https://www.openssl.org/docs/man1.0.2/man1/openssl-s_client.html)
+
+View certificates and ACLs
+
+```bash
+# These commands are run from one of Orc8r's application containers/pods
```
Comment (karthiksubraveti) [inline docs/readmes/orc8r/dev_security.md:139]: Can it be any of the pods or only Orchestrator pod. If only Orchestrator pod, can we provide command to exec into that.

File: docs/readmes/orc8r/dev_security.md [hunk 9]
```diff
@@ -0,0 +1,235 @@
...
+
+# List all registered operators, along with their ACLs and certificate serial numbers
+envdir /var/opt/magma/envdir /var/opt/magma/bin/accessc list
+
+# List all registered certificates, along with their associated identities
+envdir /var/opt/magma/envdir /var/opt/magma/bin/accessc list-certs
+```
+
+Debug northbound interface (REST API)
```
Comment (karthiksubraveti) [inline docs/readmes/orc8r/dev_security.md:148]: Can you provide may be a one liner on what each of these debug command mean and what they help identify ?

**Review by karthiksubraveti — COMMENTED** [review]
This is really good. Thanks for these changes @hcgatewood . This documentation on certs was very much needed and not available until now. Left a few comments here.

File: orc8r/tools/scripts/consolidate_protos.sh (hunk 1 above)
Comment (hcgatewood) [reply orc8r/tools/scripts/consolidate_protos.sh:5]: This script copies all Magma protos, and all protos in the Magma dependencies, to /tmp/. Useful for Wireshark, and generating the protoset described in these docs

File: docs/readmes/orc8r/dev_security.md (hunk 8 above)
Comment (hcgatewood) [reply docs/readmes/orc8r/dev_security.md:139]: Any application pod -- so orchestrator, state, subscriberdb, feg, etc

**Review by mpgermano — APPROVED** [review]

**Pull request merged**